src/daemon/rpmostreed-os-experimental.cxx

```diff
@@ -205,22 +206,35 @@ osexperimental_handle_download_packages (RPMOSTreeOSExperimental *interface,
   if (!rpmostree_context_download_metadata (ctx, DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB,
                                             cancellable, &local_error))
     goto out;
-
   sack = dnf_context_get_sack (rpmostree_context_get_dnf (ctx));
-
-  for (const char* const* it = queries; it && *it; it++)
+  RpmOstreeOverrideSourceKind source_kind;
+  source_name = rpmostree_parse_override_source (source, &source_kind, &local_error);
+  if (local_error)
+    goto out;
+  if (source_kind == RPM_OSTREE_OVERRIDE_SOURCE_KIND_REPO)
     {
-      auto query = static_cast<const char*> (*it);
-      g_autoptr(GPtrArray) pkglist = rpmostree_get_matching_packages (sack, query);
-
-      if (!pkglist || pkglist->len == 0)
+      g_autoptr(GPtrArray) pkglist = NULL; 
+      for (const char* const* it = queries; it && *it; it++)
         {
-          glnx_throw (&local_error, "No matching packages for \"%s\"", query);
-          goto out;
+          HyNevra nevra = NULL;
+          g_auto(HySubject) subject = hy_subject_create(static_cast<const char*> (*it));
+          hy_autoquery HyQuery query =
+            hy_subject_get_best_solution(subject, sack, NULL, &nevra, FALSE, TRUE, TRUE, TRUE, FALSE);
+          hy_query_filter (query, HY_PKG_REPONAME, HY_EQ, source_name);
+          pkglist = hy_query_run (query);
+          if (!pkglist || pkglist->len == 0)
+            {
+              glnx_throw (&local_error, "No matches for \"%s\" in repo '%s'", (char*) it, source_name);
+              goto out;
+            }
         }
       g_ptr_array_add (dnf_pkgs, g_object_ref (pkglist->pdata[0]));
     }
-
+  /* No need of else statement if no supported source because 
+   * rpmostree_parse_override_source covers it, future source kinds go here */
```

Comment:
We should probably still do the else here because new source types may be added without this function gaining support for them.